lib/lf/uscalfe/lagr_fe.h

```diff
@@ -710,6 +710,455 @@ class FeLagrangePoint : public ScalarReferenceFiniteElement<SCALAR> {
   unsigned degree_;
 };
 
+/**
+ * @headerfile lf/uscalfe/uscalfe.h
+ * @brief Quadratic Lagrangian finite element on a triangular reference element
+ *
+ * This is a specialization of ScalarReferenceFiniteElement.
+ * Refer to its documentation.
+ *
+ * The reference shape functions are combinations of the barycentric coordinate
+ * functions on the reference triangle.
+ *
+ * The first three shape functions are associated to the vertices,
+ * the other three shape functions to the edges of the triangle. There are
+ * no interior shape functions.
+ */
+template <typename SCALAR>
+class FeLagrangeO2Tria final : public ScalarReferenceFiniteElement<SCALAR> {
+ public:
+  FeLagrangeO2Tria(const FeLagrangeO2Tria&) = default;
+  FeLagrangeO2Tria(FeLagrangeO2Tria&&) noexcept = default;
+  FeLagrangeO2Tria& operator=(const FeLagrangeO2Tria&) = default;
+  FeLagrangeO2Tria& operator=(FeLagrangeO2Tria&&) noexcept = default;
+  FeLagrangeO2Tria() = default;
+  ~FeLagrangeO2Tria() override = default;
+
+  [[nodiscard]] base::RefEl RefEl() const override {
+    return base::RefEl::kTria();
+  }
+
+  [[nodiscard]] unsigned Degree() const override { return 2; }
+
+  /**
+   * @copydoc ScalarReferenceFiniteElement::NumRefShapeFunctions()
+   */
+  [[nodiscard]] size_type NumRefShapeFunctions() const override { return 6; }
+
+  /**  @brief One shape function attached to each node
+   * and one to each edge of the triangle. There are no interior
+   * shape functions.
+   * @copydoc ScalarReferenceFiniteElement::NumRefShapeFunctions(dim_t)
+   */
+  [[nodiscard]] size_type NumRefShapeFunctions(dim_t codim) const override {
+    LF_ASSERT_MSG(codim <= 2, "Illegal codim" << codim);
+    return (codim == 0) ? 0 : 1;
+  }
+
+  /**  @brief One shape function attached to each node
+   * and one to each edge of the triangle. There are no interior
+   * shape functions.
+   * @copydoc
+   * ScalarReferenceFiniteElement::NumRefShapeFunctions(dim_t,sub_idx_t)
+   */
+  [[nodiscard]] size_type NumRefShapeFunctions(
+      dim_t codim, sub_idx_t /*subidx*/) const override {
+    LF_ASSERT_MSG(codim <= 2, "Illegal codim" << codim);
+    return (codim == 0) ? 0 : 1;
+  }
+
+  /** @copydoc ScalarReferenceFiniteElement::EvalReferenceShapeFunctions() */
+  [[nodiscard]] Eigen::Matrix<SCALAR, Eigen::Dynamic, Eigen::Dynamic>
+  EvalReferenceShapeFunctions(const Eigen::MatrixXd& refcoords) const override {
+    LF_ASSERT_MSG(refcoords.rows() == 2,
+                  "Reference coordinates must be 2-vectors");
+
+    const size_type n_pts(refcoords.cols());
+    Eigen::Matrix<SCALAR, Eigen::Dynamic, Eigen::Dynamic> result(6, n_pts);
+
+    Eigen::ArrayXd x0 = refcoords.row(0).array();
```

Comment:
Using a const reference seems to be the cleaner approach in this case.  I also do not know, what Eigen or the compiler do.

Reply:
The `array()` function in Eigen returns an instance of `Eigen::ArrayWrapper` which just acts as a view on the matrix. This _should_ be a zero-cost abstraction. However, I'm not sure whether it is advisable to make it a reference, as the returned object is a temporary.

Reply:
yeah you have to be careful with eigen temporary objects, I've been burnt by this myself a few times. See also the [Eigen Documentation](http://eigen.tuxfamily.org/dox/TopicPitfalls.html) for some infos.
What you can do is try whether it works with `auto x0 = ...` (the unit tests will tell you ;). Otherwise the cleanest solution is really to replace every occurence of `x0` with `refcoords.row(0).array()`.

Reply:
Btw. I think the compiler is not able to optimize the current implementation even if all optimizations are turned on, so @hiptmair is right that this is not the best solution from a performance point of view.